lib/change_stream.js

```diff
@@ -476,96 +475,73 @@ function waitForTopologyConnected(topology, options, callback) {
+function resumeFromError({ changeStream, callback, eventEmitter }) {
+  const { cursor } = changeStream;
+  const {
+    options: { readPreference },
+    resumeOptions
+  } = cursor;
 
-    const error = new MongoError('ChangeStream is closed');
-    return typeof callback === 'function'
-      ? callback(error, null)
-      : changeStream.promiseLibrary.reject(error);
-  }
+  // stop listening to all events from old cursor
+  ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
 
-  const topology = changeStream.topology;
-  const options = changeStream.cursor.options;
-  const wireVersion = maxWireVersion(cursor.server);
+  // close internal cursor, ignore errors
+  cursor.close();
 
-  if (error) {
-    if (isResumableError(error, wireVersion) && !changeStream.attemptingResume) {
-      changeStream.attemptingResume = true;
-
-      // stop listening to all events from old cursor
-      ['data', 'close', 'end', 'error'].forEach(event =>
-        changeStream.cursor.removeAllListeners(event)
-      );
+  // throw it away
+  changeStream.cursor = null;
 
-      // close internal cursor, ignore errors
-      changeStream.cursor.close();
+  // attempt resume
+  const resumeAttempt = maybePromise(callback, cb => {
+    waitForTopologyConnected(changeStream.topology, { readPreference }, err => {
+      if (err) {
+        if (eventEmitter) {
+          changeStream.emit('error', err);
+          changeStream.emit('close');
+        }
+        return cb(err, null);
+      }
+      createChangeStreamCursor(changeStream, resumeOptions);
 
-      // attempt recreating the cursor
       if (eventEmitter) {
-        waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
-          if (err) {
-            changeStream.emit('error', err);
-            changeStream.emit('close');
-            return;
-          }
-          changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
-        });
-
-        return;
+        cb();
+      } else {
+        changeStream.next(cb);
       }
+    });
+  });
+  if (eventEmitter) return;
+  return resumeAttempt;
+}
 
-      if (callback) {
-        waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
-          if (err) return callback(err, null);
-
-          changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
-          changeStream.next(callback);
-        });
+// Handle new change events. This method brings together the routes from the callback, event emitter, and promise ways of using ChangeStream.
+function processNewChange({ changeStream, error, change, callback, eventEmitter }) {
+  const cursor = changeStream.cursor;
 
-        return;
-      }
+  // If the cursor is null, then change stream is attempting to resume from an error and shouldn't process a change.
+  if (!cursor) {
+    return;
```

Comment:
I think this is a behavioral change, unless you are doing the check for closed in the call for `hasNext` or `next`? It used to be that if the user was using the change stream as an iterator/cursor then this would return an error that the stream is closed. I know this is precisely the error case you are trying to solve with both of these PRs, but until we have a "request queue" I don't think we can make incremental changes that break behavior. Is this change perhaps why the test below needs to be skipped now?